sql: introduce MutableTableDescriptor struct

eriktrinh opened the pull request:
Replace the type alias which made MutableTableDescriptor and
TableDescriptor equivalent types with a struct that embeds
TableDescriptor. This forces any write of a table descriptor for schema
changes to use this struct.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/31794)
<!-- Reviewable:end -->

Commit:
sql: introduce MutableTableDescriptor struct

Replace the type alias which made MutableTableDescriptor and
TableDescriptor equivalent types with a struct that embeds
TableDescriptor. This forces any write of a table descriptor for schema
changes to use this struct.

Release note: None

vivekmenezes reviewed (changes requested):
I'm very impressed with the work you've put together here by introducing this new type. It's going to help the system get more robust. I think that the second commit is not relevant to this PR and should be removed from the PR. You can send it to me as a separate PR.

Thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-:-LPfaIHjERf9BgyYOzzh:b-ajk6fq)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/create_view.go, line 161 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-LPfWBo52V8PAZgBSR15:-LPfWBo64La1qjU5acWO:b-586xlg) ([raw file](https://github.com/cockroachdb/cockroach/blob/29ecce9e6fae0012ff74050a15f8f3db10bab0eb/pkg/sql/create_view.go#L161)):*
> ```Go
> 		if backRefMutable == nil {
> 			backRefMutable = &sqlbase.MutableTableDescriptor{TableDescriptor: updated.desc}
> 		}
> ```

Why do you need to lookup the uncommitted tables and then create a new MutableTableDescriptor here?
___
*[pkg/sql/descriptor.go, line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-LPfWbyACD1xrq9ChZGx:-LPfWbyACD1xrq9ChZGy:btpga7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/29ecce9e6fae0012ff74050a15f8f3db10bab0eb/pkg/sql/descriptor.go#L132)):*
> ```Go
> 	} else {
> 		mutDesc, ok = descriptor.(*sqlbase.MutableTableDescriptor)
> 	}
> ```

The above code seems odd
___
*[pkg/sql/physical_schema_accessors.go, line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-LPfX67U6ei8g2hjdVnF:-LPfX67U6ei8g2hjdVnG:b-2h3k6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/29ecce9e6fae0012ff74050a15f8f3db10bab0eb/pkg/sql/physical_schema_accessors.go#L146)):*
> ```Go
> 	}
> 	return nil, nil, nil
> }
> ```

Why is the change made here required for this PR?
___
*[pkg/sql/resolver.go, line 112 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-LPfXP186HvnDE4wWk9w:-LPfXP186HvnDE4wWk9x:bdzhy8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/29ecce9e6fae0012ff74050a15f8f3db10bab0eb/pkg/sql/resolver.go#L112)):*
> ```Go
> 	}
> 	return desc.(*MutableTableDescriptor), nil
> }
> ```

great! Love it!
___
*[pkg/sql/resolver.go, line 155 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-LPfXyF1ARuT19W-0Tb5:-LPfXyF1ARuT19W-0Tb6:b-jppbij) ([raw file](https://github.com/cockroachdb/cockroach/blob/29ecce9e6fae0012ff74050a15f8f3db10bab0eb/pkg/sql/resolver.go#L155)):*
> ```Go
> 		return &MutableTableDescriptor{
> 			TableDescriptor: descI.(*TableDescriptor),
> 		}, nil
> ```

You probably want to add a TODO here to remove this in a subsequent PR.
I actually feel more comfortable using a function NewMutableTableDescriptor() so we can easily track where new mutable table descriptors are created.
___
*[pkg/sql/schema_accessors.go, line 51 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-LPfYtA76ENoSX6eZuXL:-LPfYtA76ENoSX6eZuXM:bnq95dc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ee3e10883c716c0831af6eefb729afc85dadbe4/pkg/sql/schema_accessors.go#L51)):*
> ```Go
> 	// interface definitions below more intuitive.
> 	ObjectDescriptor = sqlbase.TableDescriptor
> 	// MutableTableDescriptor is provided for convenience and to make the
> ```

excellent!
___
*[pkg/sql/schema_accessors.go, line 84 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-LPfZ4eBByO2a2oTvSo4:-LPfZ4eBByO2a2oTvSo5:b-i1r8uv) ([raw file](https://github.com/cockroachdb/cockroach/blob/29ecce9e6fae0012ff74050a15f8f3db10bab0eb/pkg/sql/schema_accessors.go#L84)):*
> ```Go
> 	// IsSequence returns true if the descriptor actually describes a
> 	// Sequence resource rather than a Table.
> 	IsSequence() bool
> ```

It does appear that these ought to be methods on TableDescriptor. So users would do obj.TableDesc().IsTable()


<!-- Sent from Reviewable.io -->

eriktrinh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31794)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/create_view.go, line 161 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-LPfWBo52V8PAZgBSR15:-LPgO_bo9ZvJ0PHW4b0W:b2wis8y) ([raw file](https://github.com/cockroachdb/cockroach/blob/29ecce9e6fae0012ff74050a15f8f3db10bab0eb/pkg/sql/create_view.go#L161)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Why do you need to lookup the uncommitted tables and then create a new MutableTableDescriptor here?
</blockquote></details>

We look up the uncommitted tables because the referenced table descriptors are updated and written, which should happen on the MutableTableDescriptor as well.
___
*[pkg/sql/descriptor.go, line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-LPfWbyACD1xrq9ChZGx:-LPh46evA7AFfOot-YPj:bqzfdrm) ([raw file](https://github.com/cockroachdb/cockroach/blob/29ecce9e6fae0012ff74050a15f8f3db10bab0eb/pkg/sql/descriptor.go#L132)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

The above code seems odd
</blockquote></details>

Done. Hopefully this is more readable.
___
*[pkg/sql/physical_schema_accessors.go, line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-LPfX67U6ei8g2hjdVnF:-LPft8rs7gpEGQfed0pg:b-qs1hjf) ([raw file](https://github.com/cockroachdb/cockroach/blob/29ecce9e6fae0012ff74050a15f8f3db10bab0eb/pkg/sql/physical_schema_accessors.go#L146)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Why is the change made here required for this PR?
</blockquote></details>

We don't want to assign `nil` to `desc` in the old code and return it because the returned ObjectDescriptor will be a non-nil interface value with an underlying `nil` pointer.
___
*[pkg/sql/resolver.go, line 112 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-LPfXP186HvnDE4wWk9w:-LPgONez7wFY3ZKshiRm:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/29ecce9e6fae0012ff74050a15f8f3db10bab0eb/pkg/sql/resolver.go#L112)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

great! Love it!
</blockquote></details>

Ack.
___
*[pkg/sql/resolver.go, line 155 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-LPfXyF1ARuT19W-0Tb5:-LPftOAs0OodkotuFBZY:b-rzw1th) ([raw file](https://github.com/cockroachdb/cockroach/blob/29ecce9e6fae0012ff74050a15f8f3db10bab0eb/pkg/sql/resolver.go#L155)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

You probably want to add a TODO here to remove this in a subsequent PR.
I actually feel more comfortable using a function NewMutableTableDescriptor() so we can easily track where new mutable table descriptors are created.
</blockquote></details>

I think it would be worth it to add a second commit to this PR which does so.
___
*[pkg/sql/schema_accessors.go, line 51 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-LPfYtA76ENoSX6eZuXL:-LPgOQV0E5gdq3DTkSQT:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ee3e10883c716c0831af6eefb729afc85dadbe4/pkg/sql/schema_accessors.go#L51)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

excellent!
</blockquote></details>

Ack.
___
*[pkg/sql/schema_accessors.go, line 84 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-LPfZ4eBByO2a2oTvSo4:-LPgNlmY9bdIpoMLcAJy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/29ecce9e6fae0012ff74050a15f8f3db10bab0eb/pkg/sql/schema_accessors.go#L84)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

It does appear that these ought to be methods on TableDescriptor. So users would do obj.TableDesc().IsTable()
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

vivekmenezes reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31794)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/descriptor.go, line 129 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-LPkUSxw0MM83rWc4qMI:-LPkUSxx7yeFIniuDp19:bb9qtqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/5beaebd0f00e29454e61e9a70ebfc144d289cffd/pkg/sql/descriptor.go#L129)):*
> ```Go
> 	var mutDesc *sqlbase.MutableTableDescriptor
> 	switch d := descriptor.(type) {
> 	case *sqlbase.MutableTableDescriptor:
> ```

how can a descriptor.(type) ever be a MutableTableDescriptor?
___
*[pkg/sql/resolver.go, line 155 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-LPfXyF1ARuT19W-0Tb5:-LPkVVno2zD3tjU9MTqQ:bfbl86a) ([raw file](https://github.com/cockroachdb/cockroach/blob/29ecce9e6fae0012ff74050a15f8f3db10bab0eb/pkg/sql/resolver.go#L155)):*
<details><summary><i>Previously, eriktrinh (Erik Trinh) wrote…</i></summary><blockquote>

I think it would be worth it to add a second commit to this PR which does so.
</blockquote></details>

this looks fine for now


<!-- Sent from Reviewable.io -->

vivekmenezes reviewed (commented):
This looks good. Feel free to incorporate the feedback and merge. Thanks for making it happen!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-:-LPkVt5x2xK-xxnc_3Kb:b-rj864n)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

vivekmenezes reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-:-LPkW0P371MR6uyuTYaP:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

eriktrinh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31794)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/descriptor.go, line 129 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31794#-LPkUSxw0MM83rWc4qMI:-LPlVfdd2EoMEC-SW1gg:b-266k4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/5beaebd0f00e29454e61e9a70ebfc144d289cffd/pkg/sql/descriptor.go#L129)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

how can a descriptor.(type) ever be a MutableTableDescriptor?
</blockquote></details>

this is called in create_table and create_view where it uses created MutableTableDescriptors which implements DescriptorProto


<!-- Sent from Reviewable.io -->

eriktrinh commented:
bors r+
